### PR TITLE
Introducing mypy checks

### PR DESCRIPTION
**Pull request by Tacire**
**What**:

Update typing and introducing automatic mypy checks on PR

**Why**:

To increase code stability and readability

**How**:

Fixed some typing issues and created an mypy workflow that runs on pull requests

**Checklist**:

- [x] Tests
- [x] PR merge commit message adjusted
- [ ] Documentation

### CODE REVIEW

File: autohooks/precommit/run.py [hunk 1]
```diff
@@ -141,7 +144,7 @@ def init(self, total: int) -> None:
         """
         self._progress.update(self._task_id, total=total)
 
-    def update(self, advance: Optional[int] = 1) -> None:
+    def update(self, advance: int = 1) -> None:
```
Comment (Tacire) [inline autohooks/precommit/run.py:147]: I'm a bit unsure about changing the type of an parameter.  But since _progress.advance() needs an fixed float and update() already sets a default value for advance, i believe this might be the cleanest solution to fix an occuring mypy error (and the Optional[int] might simply be not necessary)

File: autohooks/precommit/run.py [hunk 2]
```diff
@@ -129,7 +132,7 @@ class ReportProgress:
 
     def __init__(self, progress: Progress, task_id: int) -> None:
         self._progress = progress
-        self._task_id = task_id
+        self._task_id = TaskID(task_id)
```
Comment (Tacire) [inline autohooks/precommit/run.py:135]: Since the task_id needs to be of the NewType "TaskID" and i am hesitant to touch the type of an parameter, i'm simply converting the int to the newtype TaskID (which is basically also an int)

File: autohooks/cli/plugins.py [hunk 3]
```diff
@@ -73,11 +73,11 @@ def add_plugins(term: Terminal, args: Namespace) -> None:
 
     if config.has_autohooks_config():
         settings = config.settings
-        existing_plugins = set(settings.pre_commit)
+        existing_plugins = set(settings.pre_commit)  # type: ignore
```
Comment (Tacire) [inline autohooks/cli/plugins.py:76]: in multiple places throughout the code i placed some "# type: ignore" to ignore mypy errors that occur because "settings" might not be definied. But since we checked earlier if config has an autohooks config (line 74), this should be not an error at all. So i think these errors can simply be ignored. If you believe there might be an easier solution please share it with me.

Comment (Tacire) [reply autohooks/cli/plugins.py:76]: (or a cleaner solution)

File: autohooks/settings.py [hunk 4]
```diff
@@ -70,7 +70,9 @@ def write(self, filename: Path) -> None:
         overridden.
         """
         if filename.exists():
-            toml_doc = tomlkit.loads(filename.read_text())
+            toml_doc: Union[
```
Comment (Tacire) [inline autohooks/settings.py:73]: Some errors occure because of this typing:

`autohooks/settings.py:81: error: Unsupported right operand type for in ("Union[Item, Container]")  [operator]
autohooks/settings.py:82: error: Unsupported target for indexed assignment ("Union[Item, Container]")  [index]
autohooks/settings.py:89: error: Value of type "Union[Item, Container]" is not indexable  [index]
autohooks/settings.py:89: error: Item "Item" of "Union[Any, Item, Container]" has no attribute "update"  [union-attr]`

So there seems to be an problem with the typing Union[TOMLDocument, Table]. Might anyone be able to give me some insights into what might be missing?

File: autohooks/terminal.py [hunk 5]
```diff
@@ -36,7 +36,7 @@
     "warning",
 )
 
-__term = None  # pylint: disable=invalid-name
+__term = Terminal()
```
Comment (Tacire) [inline autohooks/terminal.py:39]: This Line was added to remove the possibility of an undefined _term (which was bothering mypy).
Now the following error occures:
`autohooks/terminal.py:113: error: "RichTerminal" has no attribute "print_overwrite"  [attr-defined]`
Which indeed seems to be true (at least i couldn't find this attribute). Any thoughts on this?

File: autohooks/precommit/run.py (hunk 1 above)
Comment (Tacire) [reply autohooks/precommit/run.py:147]: Furthermore mypy has some more problems with this file:
`autohooks/precommit/run.py:200: error: Argument 1 to "ReportProgress" has incompatible type "rich.progress.Progress"; expected "autohooks.terminal.Progress"  [arg-type]`
`autohooks/precommit/run.py:213: error: "Progress" has no attribute "finish_task"  [attr-defined]`
I still strugle to understand the context in which these errors occure (I understand that Progress indeed does not have an attribute finish_task, but can't figure out why it even should have it). Some further insights would be really appreciated :)

File: autohooks/config.py [hunk 6]
```diff
@@ -84,11 +84,11 @@ def is_empty(self) -> bool:
         return not bool(self._config_dict)
 
 
-def _gather_mode(mode: Optional[str]) -> Mode:
+def _gather_mode(mode_string: Optional[str]) -> Mode:
     """
     Gather the mode from a mode string
     """
-    mode = Mode.from_string(mode)
+    mode = Mode.from_string(mode_string)
```
Comment (bjoernricks) [inline autohooks/config.py:91]: Not sure why this change is necessary. Personally I don't like adding a type to a variable name. Would like to revert this change.

File: .github/workflows/ci-python.yml [hunk 7]
```diff
@@ -24,6 +24,23 @@ jobs:
...
+  type-check:
+    name: Type-check
+    runs-on: "ubuntu-latest"
+    strategy:
+      matrix:
+        python-version:
+          - 3.7
+          - 3.8
+          - 3.9
+          - "3.10"
```
Comment (bjoernricks) [inline .github/workflows/ci-python.yml:36]:
```suggestion
          - "3.10"
          - "3.11"
```

File: autohooks/precommit/run.py (hunk 2 above)
Comment (bjoernricks) [reply autohooks/precommit/run.py:135]: Personally I don't like this kind of typing using [NewType](https://docs.python.org/3/library/typing.html#newtype) but for removing a mypy issue this is fine.

File: autohooks/precommit/run.py (hunk 1 above)
Comment (bjoernricks) [reply autohooks/precommit/run.py:147]: For me both versions of the update method are correct (because the called method has a default argument of 1 too) therefore the change is fine to fix a mypy issue.

File: autohooks/settings.py (hunk 4 above)
Comment (bjoernricks) [reply autohooks/settings.py:73]: I don't understand this part. IMHO it must be `toml_doc: tomlkit.TOMLDocument`

File: autohooks/terminal.py (hunk 5 above)
Comment (bjoernricks) [reply autohooks/terminal.py:39]: I am really not sure about this change. When we apply this change the `_set_terminal` function should be rewritten too because it doesn't make sense to keep it as it is anymore. Personally I am in favor of not applying this change.

File: autohooks/utils.py [hunk 8]
```diff
@@ -49,7 +51,7 @@ def exec_git(*args: str, ignore_errors: bool = False) -> str:
         exec_git("commit", "-m", "A new commit")
     """
     try:
-        cmd_args = ["git"]
+        cmd_args: List[Union[PathLike, str]] = ["git"]
```
Comment (bjoernricks) [inline autohooks/utils.py:54]: Not sure if this is correct. Shouldn't it be `List[str]` only? If `Union[PathLike, str]` is valid could you check if there is a test case and if not could you add one?

File: autohooks/config.py (hunk 6 above)
Comment (Tacire) [reply autohooks/config.py:91]: this is necessary because mode is defined as an Optional[str] but later overwriten with a "Mode"-Object. That's a problem for mypy. The name is inspired by the parameter of "from_string", that is also called modestring

File: autohooks/utils.py (hunk 8 above)
Comment (Tacire) [reply autohooks/utils.py:54]: i will check these changes again and come back to you. I think some of these changes where made in our joint coding session to eliminate other errors.

Comment (Tacire) [reply autohooks/utils.py:54]: `utils.py` cmd_args is typed as List[Union[PathLike, str]] because in the next line cmd_args.extend(args) is called. If you activate mypy and observe correct typing, the extend-method only accepts parameters of an type that is already in the List. And since an PathLike-Object is given as parameter for *args in line 69 (in get_git_directory_path()) cmd_args has to be prepared with the correct typing. 
`def extend(self, __iterable: Iterable[_T]) -> None: ...`
I realise that this typing is pretty confusing. So maybe an easier way is simply turning the PathLike-Object into an string early on and remove any PathLike parameters. I'm a bit unsure about properties that are lost in this process. I will check in what way this is covered in the test cases.

`terminal.py` Since the code doesn't prohibit the usage of terminal-functions before set_terminal is called i believe the only other typing solution would be to set some flags for mypy to ignore the lines / errors. Another option would be to define the behavior for each time "_term" is used when it is still "None". And of course i can update _set_terminal if we stay with this typing solution. Let me know what you would prefer (or if i missed anything)

File: autohooks/settings.py (hunk 4 above)
Comment (Tacire) [reply autohooks/settings.py:73]: Your right. Thats the correct typing. That was an error on my side. The difficult part is that we are calling entries of the TOMLDocument (for example) "tool" (that should be a table) and check if something is "in" it. later we try to call toml_doc["tool"]["autohooks"].update(). In the context of this module "tool" doesn't necessary need to be defined, so mypy is clueless what it could be. Sadly i don't think i can type "tool" in here. Or do you maybe have an idea how i could do that? So my options seem to be to ignore the errors or check the instance of ["tool"] and ["autohooks"] each time we use them.

File: autohooks/config.py (hunk 6 above)
Comment (Tacire) [reply autohooks/config.py:91]: Do you have a cleaner solution in mind? In my opinion that's the best way to go (including the variable name)

File: autohooks/utils.py (hunk 8 above)
Comment (Tacire) [reply autohooks/utils.py:54]: > I realise that this typing is pretty confusing. So maybe an easier way is simply turning the PathLike-Object into an string early on and remove any PathLike parameters. I'm a bit unsure about properties that are lost in this process. I will check in what way this is covered in the test cases.

In `test_utils.py` the PathLike-Object seems to get tested in exactly that way. So i think this change is convenient and might be even better tested
```
    def test_with_env_pwd(self):
        with TemporaryDirectory() as f:
            temp_path = Path(f)

            exec_git("-C", str(temp_path), "init")

            os.chdir(str(temp_path))

            git_dir_path = (temp_path / ".git").resolve()

            git_hook_dir_path = get_git_hook_directory_path()
            self.assertEqual(git_hook_dir_path, git_dir_path / "hooks")
```

File: autohooks/config.py (hunk 6 above)
Comment (bjoernricks) [reply autohooks/config.py:91]: Ok should be fine then :-)

File: autohooks/cli/plugins.py (hunk 3 above)
Comment (bjoernricks) [reply autohooks/cli/plugins.py:76]: Should be find to just add an ignore statement here.

File: autohooks/settings.py (hunk 4 above)
Comment (bjoernricks) [reply autohooks/settings.py:73]: I am really not sure what's the problem. At the end `TOMLDocument` is also a dict and implements the [__contains__](https://docs.python.org/3.10/reference/datamodel.html#object.__contains__) and [__getitem__](https://docs.python.org/3.10/reference/datamodel.html#object.__getitem__) protocols required for both.

File: autohooks/settings.py [hunk 9]
```diff
@@ -70,9 +70,9 @@ def write(self, filename: Path) -> None:
         overridden.
         """
         if filename.exists():
-            toml_doc: Union[
-                tomlkit.toml_document.TOMLDocument, tomlkit.items.Table
-            ] = tomlkit.loads(filename.read_text())
+            toml_doc: tomlkit.toml_document.TOMLDocument = tomlkit.loads(
```
Comment (bjoernricks) [inline autohooks/settings.py:73]: You can shorten the typing by using the main module
```suggestion
            toml_doc: tomlkit.TOMLDocument = tomlkit.loads(
```

File: autohooks/settings.py (hunk 4 above)
Comment (Tacire) [reply autohooks/settings.py]: TOMLDocument is not the problem but rather the entries that are undefined (in terms of typing). At least that's what i'm interpreting in these errors:
```
autohooks/settings.py:81: error: Unsupported right operand type for in ("Union[Item, Container]")  [operator]
autohooks/settings.py:82: error: Unsupported target for indexed assignment ("Union[Item, Container]")  [index]
autohooks/settings.py:89: error: Value of type "Union[Item, Container]" is not indexable  [index]
autohooks/settings.py:89: error: Item "Item" of "Union[Any, Item, Container]" has no attribute "update"  [union-attr]
```
And i don't think that i can type them in this modul. So my solution would be to ignore these errors if there is no other way

```
  if filename.exists():
            toml_doc: tomlkit.toml_document.TOMLDocument = tomlkit.loads(
                filename.read_text()
            )
        else:
            toml_doc = tomlkit.document()

        if "tool" not in toml_doc:
            toml_doc["tool"] = tomlkit.table(is_super_table=True)
81      if "autohooks" not in toml_doc["tool"]:
82          toml_doc["tool"]["autohooks"] = tomlkit.table()

        config_dict = {
            "mode": str(self.mode.get_effective_mode()),
            "pre-commit": sorted(self.pre_commit),
        }

89      toml_doc["tool"]["autohooks"].update(config_dict)

```
